substrate/frame/elections-phragmen/src/lib.rs

```diff
@@ -591,13 +591,16 @@ pub mod pallet {
 		/// ## Complexity
 		/// - Check is_defunct_voter() details.
 		#[pallet::call_index(5)]
-		#[pallet::weight(T::WeightInfo::clean_defunct_voters(*_num_voters, *_num_defunct))]
+		#[pallet::weight(T::WeightInfo::clean_defunct_voters(*num_voters, *num_defunct))]
 		pub fn clean_defunct_voters(
 			origin: OriginFor<T>,
-			_num_voters: u32,
-			_num_defunct: u32,
+			num_voters: u32,
+			num_defunct: u32,
 		) -> DispatchResult {
 			let _ = ensure_root(origin)?;
+			// We don't check the weight witness since it is a root call.
+			let _ = (num_voters, num_defunct);
```

Comment:
But users could add more defunct voters by registering, voting for themselves, and then renouncing candidacy. The `iter` below will still go through everything, and the weight will be wrong. Being a Root call it would probably be in `on_initialize` and thus mandatory.

I think it would need to drain, return an `Option<cursor>`, and schedule another cleanup attempt in a future block if there were any left over?

Reply:
I fixed it with this diff now:  
```patch
                        <Voting<T>>::iter()
+                               .take(num_voters)
                                .filter(|(_, x)| Self::is_defunct_voter(&x.votes))
+                               .take(num_defunct)
                                .for_each(|(dv, _)| Self::do_remove_voter(&dv));
```

which should ensure that the weight cannot be front-run. Thanks for pointing it out!